### ISSUE TITLE
Package Pre Release Versioning

### ISSUE THREAD

**Issue by skorfmann** [hashicorp/terraform-cdk#561]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Each commit on the `main` branch gets automatically released as pre-release version to all targets: npm / pypi / nuget / maven

Currently, we take the current version number (e.g. `0.1.0` as of today) and append a `pre-<BUILD_NUMBER` in all cases but pypi, where it's `dev.<BUILD_NUMBER>`. This results in something like `0.1.0-pre-178`, which is a bit misleading from my point of view. I think I'd expect the pre-release to indicate the next version. So, something like `0.2.0-pre-178`.

### Overview

| Target | release version | current pre-release | actually expected pre-release      |
|--------|-----------------|---------------------|---------------|
| [NPM](https://www.npmjs.com/package/cdktf)    | 0.1.0           | 0.1.0-pre.188       | 0.2.0-pre.188 |
| [pypi](https://pypi.org/project/cdktf/0.1.0/#history)   | 0.1.0           | 0.1.0.dev188        | 0.2.0.dev188  |
| [nuget](https://www.nuget.org/packages/HashiCorp.Cdktf)  | 0.1.0           | 0.1.0-pre.188       | 0.2.0-pre.188  |
| [maven](https://search.maven.org/artifact/com.hashicorp/cdktf)  | 0.1.0           | 0.1.0-pre.188       | 0.2.0-pre.188  |

### Goals

The goals I see:

- A pre-release version should indicate that it's a snapshot of unreleased contents of the `main` branch
- A pre-release version should track the next version compared to the current "official" release
- For the normal workflow, I'd expect the next minor version as pre-release target. I think we can ignore patch versions for next releases. 
- Major version pre-releases will be handled manually somehow (we'll deal with that when we have the problem)
- Package managers should ignore pre-releases by default and only install them by explicit request
- It should work with the `jsii-pacmak` tooling, which may map versions from NPM to the other targets (in particular for [Python](https://github.com/aws/jsii/blob/47a40f15195c00368c55bb9d08eded063d970dec/gh-pages/content/user-guides/lib-author/configuration/targets/python.md#prerelease-versions))

### Release Workflow

The current [release workflows](https://github.com/hashicorp/terraform-cdk/blob/1c5a39fd59c3e0101b3f577589320fa81ea2d4cb/.github/workflows/release.yml) are trying to publish the currently specified version on reach run for the `main` branch. See [here](https://github.com/hashicorp/terraform-cdk/runs/2498985590?check_suite_focus=true) for example:

```
npm notice 
npm ERR! code E403
npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/cdktf - You cannot publish over the previously published versions: 0.3.0.
npm ERR! 403 In most cases, you or one of your dependencies are requesting
npm ERR! 403 a package version that is forbidden by your security policy.
```

That's true for all registry targets. We should keep this in mind, so we're not accidentally publishing new releases. 

The relevant [pre release workflows](https://github.com/hashicorp/terraform-cdk/blob/main/.github/workflows/release_next.yml) are separate, but are using the same [source](https://github.com/hashicorp/terraform-cdk/blob/1c5a39fd59c3e0101b3f577589320fa81ea2d4cb/tools/align-version.sh) for the version. 

### Versioning References

More context for versioning for the different targets:

- [nuget](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages)
- [pypi](https://www.python.org/dev/peps/pep-0440/)
- [maven](https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN401)

**Comment by jsteinich** [comment]
For maven, using `SNAPSHOT` is probably the most standard, but it would require some changes relative to other packages. Alpha and beta are known qualifiers, so they could be good alternatives, especially since that's probably the standard for nuget. [This site](https://octopus.com/blog/maven-versioning-explained) gives a pretty thorough breakdown.

**Comment by skorfmann** [comment]
> For maven, using `SNAPSHOT` is probably the most standard, but it would require some changes relative to other packages.

Which packages do you mean?

**Comment by skorfmann** [comment]
> For maven, using `SNAPSHOT` is probably the most standard, but it would require some changes relative to other packages. Alpha and beta are known qualifiers, so they could be good alternatives, especially since that's probably the standard for nuget. 

It looks like the `pre` suffix is good enough though. Nuget will treat any suffix as pre-release version (in reverse alphabetical order - see [here](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages#semantic-versioning)). 

I think it's [similar](https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN400) for Maven.

> All versions with a qualifier are older than the same version without a qualifier (release version).
> 
> For example:
> 
> 1.2-beta-2 is older than 1.2.
> 
> Identical versions with different qualifier fields are compared by using basic string comparison.
> 
> For example:
> 
> 1.2-beta-2 is newer than 1.2-alpha-6.

So, for simplicity reasons I'd just stick to the existing suffixes and just change the targeted minor version.

**Comment by jsteinich** [comment]
For maven, `SNAPSHOT` is nice because it has special semantics. While it has an exact version component (typically data +- build number), it can be used without the exact version and tools will periodically check for new snapshot versions. This is often a useful property, but given that a pre-release version of the cli may have compatibility requirements for the pre-release library, we may not want to go that route.
Using an recognized qualifier, such as `pre`, changes how tools sort versions, specifically: "Versions with unrecognized qualifiers are treated as later releases than an unqualified version, and unrecognized qualifiers are compared as case insensitive strings."

For nuget, it doesn't matter as much, but using `alpha` or `beta` would be more in line with conventions.

**Comment by DanielMSchmidt** [comment]
Using semantic release (if used in #656) could be used here as well, we would need to replace the pre / dev / SNAPSHOT identifier but that should be easy enough

**Comment by skorfmann** [comment]
> We could configure main as pre-release branch

And how would you do a normal release then?

**Comment by skorfmann** [comment]
This was implemented by @jsteinich in #1257 - thanks again!
